feat: Added card counts for columns with extra information while filtering

RARgames opened the pull request:
Please don't "fix" the "<no_cards> cards" alignment, after adding column collapsing feature it will look good :D
I'll try to implement this feature soon, together with board card counts, project card counts and improved version of dashboard to include some stats about projects.

Here is an example of how it works:
![image](https://user-images.githubusercontent.com/13639766/211877593-f3b3b79b-efba-4c7e-93e6-577969e23e6c.png)
and after filtering:
![image](https://user-images.githubusercontent.com/13639766/211877678-9786acb0-aa18-43cf-b70b-aba2ed60ad50.png)

meltyshev commented:
Hi and thank you for your contribution!

I have a few fixes that I would make before accepting the PR:

1. Move `cardModelsFull` and `isFiltered` from `getFilteredOrderedCardsModelArray` to separate selectors and call all 3 selectors in the container. This will avoid modifying the files where the `getFilteredOrderedCardsModelArray` function is called and also this function will not be responsible for 3 things at once.
2. In `cardsCountText` we have to use [plurals](https://www.i18next.com/translation-function/plurals) functionality instead of checking `length === 1`, because in some languages there are multiple plurals. Also, I would translate the whole line without putting parts of it in a separate line of translation.
3. Add a setting in user preferences to be able to disable card counts if the user doesn't need it.

> Please don't "fix" the "<no_cards> cards" alignment, after adding column collapsing feature it will look good :D

I would make it on the same line as the column name, since when we add the collapsing feature we can move it back at any time.

I can make some commits with fixes, but not until the weekend or next week.

RARgames commented:
Hey,
Added card counts for board and column collapsing as I think these changes are logical together (Column collapsing depends on card counts, if we want to show counts there).
I'll take a look at you feedback in a sec and try to include it.

RARgames commented:
> 1. Move `cardModelsFull` and `isFiltered` from `getFilteredOrderedCardsModelArray` to separate selectors and call all 3 selectors in the container. This will avoid modifying the files where the `getFilteredOrderedCardsModelArray` function is called and also this function will not be responsible for 3 things at once.

Fixed this, also changed cardIds -> filteredCardIds, cardIdsFull ->cardIds (It should be better than cardIdsFull)

> I would make it on the same line as the column name, since when we add the collapsing feature we can move it back at any time.

Added column collapsing, everything seems to work fine, only #377 still looks bad.

nickbe commented:
Hi RAR, having used collapsing columns for a time myself, I must say that it's a feature that has more complications than benefits. First of all if you hide 10 out of 20 columns, you will have to go through the columns one by one - showing/collapsing - to review everything. Also if you have too many columns in a kanban board then you should seriously consider to reorganize the board and maybe split it into several others instead. 

We are working on a better solution already. When it's finished it will provide a much more elegant way to achieve something similar. 

Regarding the card counters we can add this by optimizing the layout and making it switchable in a users settings.

RARgames commented:
> Hi RAR, having used collapsing columns for a time myself, I must say that it's a feature that has more complications than benefits. First of all if you hide 10 out of 20 columns, you will have to go through the columns one by one - showing/collapsing - to review everything. Also if you have too many columns in a kanban board then you should seriously consider to reorganize the board and maybe split it into several others instead.

Cannot agree with that. There are some workflows that require column collapsing, or just benefit from it.

> We are working on a better solution already. When it's finished it will provide a much more elegant way to achieve something similar.

Would love to hear more about it. Also do you have any ETA on this and the current developments that will be available as open source later?

nickbe commented:
Can you give us a workflow example out of public usage where this is needed and beneficial?

You know how it is with ETAs. I'd rather not. But it's a very active development and we are not thinking in years now.

RARgames commented:
Sure, just to name a few:
- a company mantaitaining many programs for a platform, programs are not in active development, but sometimes need an update, a board is used just to collect the tasks in lists - 1 list per program. When the time comes someone uncollapses the list to start working on them.
- a company managing many projects, each project has its own board/s, but there are only a few new "management" tasks each month, sometimes 0 tasks (this tasks are only available for people with different access level). The management can focus only on a few projects at one time, they don't care about other projects during the meeting.
- for prototyping a project when you don't need complicated structures, only a plain view and you focus on some of the aspects only, if a project is funded, it will be moved from prototyping board to a different one/s.
- personal todo board, where you keep things you want to do, or have to do in different categories. Some of them e.g. ideas for later, specific areas that you don't want to focus right now, but you will later can be collapsed.

Making a separate board for it doesn't make sense, column collapsing does. 

There are many more scenarios like this. 

In addition this feature does not make it harder for someone that don't like column collapsing to use boards. The current state (when there is always a small triangle to collapse/uncollapse board) doesn't seem to affect people not using this feature much. I cannot see the downsides of having this as an extra, even togglable feature. 

In the above scenarios, users will probably benefit from the small addition of being able to drop cards onto collapsed columns.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/plankanban/planka?pullRequest=380) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/plankanban/planka?pullRequest=380) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/plankanban/planka?pullRequest=380) it.</sub>

nickbe commented:
Hi again, first of all it seems that you completely forgot about the labels here. Their job is to categorize and filter. Multiple of your ideas could be implemented just with labels here.

Then for us there's the problem of mobile usage. planka is not yet optimized for mobiles, but it will be. There we want to swap from column to column, but with collapsed columns this will either look strange or we have to disable it there... (but do what? Currently i cannot think how to handle these columns then.

meltyshev commented:
Hi! We've decided that this is a really needed feature and we are ready to accept your PR, but first we'll modify the visual part a little bit. But we can't do it without signing the CLA on your side, it became necessary after the recent license change. Thank you 🙏

RARgames commented:
Hey @meltyshev,
Alright!
I'm not sure what is the license change about, but I'll read it and sign the CLA.

Also, I can see that my vision for boards may differ at some points from Planka's vision. I need some features that you may find unnecessary. For my project, I don't really need a complicated license. That's why I'll continue to add features that I need in my repo. My repo is still on the MIT license. Feel free to use my additions in Planka if you want.

Can I directly implement your fixes (commits with a fix prefix) in my version of boards under the MIT license?

meltyshev commented:
Hey! It's a good decision to make your own version, it'll be faster and easier for you instead of constantly syncing with us 👍

> Can I directly implement your fixes (commits with a fix prefix) in my version of boards under the MIT license?

It seems that using code with the AGPL license in the MIT project is not forbidden, but then the entire project must change its license to AGPL. Honestly, I don't know how it works when it comes to small changes.

RARgames commented:
Sorry, I started reading the new license terms and forgot about this.

> It seems that using code with the AGPL license in the MIT project is not forbidden, but then the entire project must change its license to AGPL. Honestly, I don't know how it works when it comes to small changes.

If you just confirm that it's ok with you (me using fixes under the old license) that's enough for me.
I just want to keep it simple with license that I fully understand.
